[2단계 - 웹 자동차 경주] 아마란스(송세연) 미션 제출합니다.

amaran-th opened the pull request:
안녕하세요 춘식! 2단계도 잘 부탁드립니다!

Commit:
docs : README에 2단계 미션 요구사항 추가 및 흐름도 업데이트

Commit:
feat : DTO 네이밍 변경 및 DB에서 승자 불러오는 로직 구현

Commit:
feat : 이력 조회 기능 구현

Commit:
feat : 콘솔 어플리케이션 출력 양식변경

Commit:
feat : 콘솔 애플리케이션/웹 애플리케이션의 공통 기능을 별도 도메인 기능으로 분리

Commit:
fix : RacingRequest 관련 오류 해결

- 기본 생성자 추가하여 오류 해결

Commit:
docs : 클래스다이어그램 제거

Commit:
chore : 불필요한 코드 제거

bimppap commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -15,6 +14,10 @@ private Car(CarName name) {
         this.name = name;
         this.position = Position.of(ZERO_POSITION);
     }
+    public Car(CarName name, Position position) {
```
[생성자 체인 기법](https://www.geeksforgeeks.org/constructor-chaining-java-examples/)을 적용해봐도 좋을 것 같습니다!

amaran-th replied:
반영했습니다! 감사합니다:)

bimppap commented on src/main/java/racingcar/dto/RacingEntity.java:
```diff
@@ -0,0 +1,28 @@
+
+import java.sql.Timestamp;
+
+public class RacingEntity {
```
이건 사용하는 곳이 없는 것 같아요 👀

amaran-th replied:
삭제했습니다!

bimppap commented on src/main/java/racingcar/exception/Advice.java:
```diff
@@ -11,13 +11,13 @@ public class Advice {
-    public ExceptionMessageDTO handleIllegalInput(IllegalArgumentException exception) {
-        return new ExceptionMessageDTO(ILLEGAL_INPUT_ERROR, exception.getMessage());
+    public ExceptionResponse handleIllegalInput(IllegalArgumentException exception) {
+        return new ExceptionResponse(ILLEGAL_INPUT_ERROR, exception.getMessage());
```
ExceptionResponse 의 type 필드 들어가는 숫자는 무슨 기준으로 정한 건지 궁금합니다 👀

amaran-th replied:
숫자는 임의로 결정했습니다! 입력 에러는 20, 그 외 에러는 -1입니다.
프론트에서 type 숫자를 기준으로 에러를 처리할 수 있게 하기 위해 추가했습니다.

bimppap commented on src/main/java/racingcar/dao/QueryingDAO.java:
```diff
@@ -0,0 +1,42 @@
+import racingcar.utils.RandomNumberGenerator;
+
+@Repository
+public class QueryingDAO {
```
이런 식의 네이밍은 추상적이라 어떤 목적의 DAO인지 불분명해보입니다. 🥲InsertingDao처럼, 기능 위주로 분리하면서 Querying을 사용하게 된 것처럼 보이네요. 
1단계에서 가볍게 짚었지만 DAO(또는 레포지토리)는 도메인 혹은 업무 단위로 분리하는게 좋습니다. 또한 데이터 모델은 아래 명명 법칙을 따르는 게 일반적입니다.

- 현업업무에서 사용하는 용어를 사용한다.
- 약어를 사용하지 않는다.
- 단수명사를 사용한다.
- 모든 엔터티에서 유일하게 이름이 부여되어야 한다.
- 엔터티 생성의미대로 이름을 부여한다.

관련해서 스택오버플로우에 [참고하면 좋은 글](https://stackoverflow.com/questions/4702728/relational-table-naming-convention/4703155#4703155)도 있으니 한 번 읽어보면 좋을 것 같습니다. 🙂

amaran-th replied:
반영해서 RacingDAO라는 하나의 DAO 클래스로 만들었습니다!😊

bimppap commented on src/main/java/racingcar/controller/ConsoleController.java:
```diff
@@ -19,39 +22,33 @@ public ConsoleController(OutputView outputView, InputView inputView) {
         try {
-            return Trial.of(Converter.convertStringToInt(inputView.getTrial()));
+            String input = inputView.getTrial();
+            Trial.of(Converter.convertStringToInt(input));
```
이건 이제 사용되지 않는 코드군요 😎

amaran-th replied:
앗 그건 입력 값에 대한 검증을 해줘야해서 넣었습니다! 저게 없으면 각 입력값마다 검증을 하는 게 아니라 carNames와 trialCount 값까지 다 받은 후에 검증을 하게 되어서요...!
유효성 검증을 inputView가 아니라 도메인 객체가 하고 있어서 이렇게 했는데 다른 방법이 떠오르지 않더라구요ㅜ

amaran-th replied:
아래 글을 참고해서 이해한 대로 입력 받은 값을 가공하는 역할을 서비스에서 컨트롤러로 옮겼습니다! 
[블로그 글](https://blog.naver.com/zxxne/222482502102)

bimppap commented on src/main/java/racingcar/dto/CarInfoEntity.java:
```diff
@@ -1,13 +1,13 @@
     private boolean isWinner;
 
-    public CarInfoDto(int racingId, String name, int position, boolean isWinner) {
+    public CarInfoEntity(int racingId, String name, int position, boolean isWinner) {
```
Dto를 Entity로 바꾼 이유가 있을까요? Dto와 Entity는 엄연히 다른 개념인데 dto 파일에 넣게 된 이유가 궁금합니다 🤔

amaran-th replied:
DB의 데이터와 1:1 매칭되는 객체라고 판단되어서 Entity로 바꿔주었습니다!
패키지 분류를 신경쓰지 못했네요...! 둘이 다르지만 비슷한 개념이라고 생각해서...
entity 패키지를 따로 만들어서 옮겨주었습니다!

bimppap commented on src/main/java/racingcar/service/RacingService.java:
```diff
@@ -0,0 +1,43 @@
+import racingcar.dto.RacingResultResponse;
+
+@Service
+public class RacingService {
```
컨트롤러에 있던 비즈니스 로직을 서비스로 잘 분리해주셨네요 👍

bimppap commented on src/main/java/racingcar/Application.java:
```diff
@@ -1,6 +1,10 @@
 package racingcar;
```
Application 코드에 코멘트를 달 수가 없어서 여기에 답니다ㅎㅎ

ConsoleController를 service라는 변수명으로 사용하고 있는데, 지금은 서비스 클래스가 있으니 혼동이 오지 않도록 controller 또는 다른 이름으로 바꾸는 게 좋을 것 같습니다 🙂

bimppap replied:
<img width="478" alt="image" src="https://user-images.githubusercontent.com/48675973/232840523-bed3a4d1-03fa-4f63-be7d-9628eaeada85.png">

<img width="825" alt="image" src="https://user-images.githubusercontent.com/48675973/232840823-f9789d3d-6c31-45c0-84d1-ed607117bc30.png">

둘 다 수정된 코드로 잘 동작하네요 👍

amaran-th replied:
헉 수정하는 걸 잊어버렸네요
controller로 변경했습니다! 감사합니다~~

bimppap commented on src/main/java/racingcar/controller/WebController.java:
```diff
@@ -1,51 +1,34 @@
-    public Trial getTrial(String input) {
-        return Trial.of(Converter.convertStringToInt(input));
+    public RacingResultResponse run(@RequestBody RacingRequest dto) {
+        Cars updatedCars = RacingGame.run(dto);
```
변수명에 대해 한번 고민해보면 좋을 것 같습니다 🙂(콘솔컨트롤러에서도 마찬가지입니다!)

bimppap replied:
그리고 RacingGame에서 run 하는 건 비즈니스 로직이라 서비스에 있는 게 맞을 것 같습니다. 스프링의 컨트롤러와 서비스가 어떤 역할을 맡는지 정리해보면 좋을 것 같아요. (원하신다면 심화로 스프링MVC에 대해 공부해보셔도 좋습니다ㅋㅋ)

amaran-th replied:
RacingGame을 없애고 안에 있던 run 메서드를 Service로 옮겼습니다! 감사합니다

bimppap commented on src/main/java/racingcar/domain/RacingGame.java:
```diff
@@ -0,0 +1,27 @@
+import racingcar.utils.Converter;
+import racingcar.utils.RandomNumberGenerator;
+
+public class RacingGame {
```
RacingGame 메서드들이 static으로 쓰이지 않게 변경해보는 건 어떨까요?

amaran-th replied:
앞의 코멘트를 반영해서 RacingGame을 제거해줬습니다!
public static 메서드는 가능한 지양하는 것이 좋을까요??

bimppap replied:
제거 잘해주셨네요!! 👍👍 정적 팩토리 메서드와는 다르게 정적 메서드의 경우는 여러 곳에서 쓰이는 게 아니라면 객체지향성을 해쳐서 지양하는 것이 좋습니다 🙂 여러 곳에서 쓰인다면 utils 와 같은 이름의 패키지에서 관리를 해주는 게 좋습니다.

bimppap commented on src/main/java/racingcar/dto/RacingResultResponse.java:
```diff
@@ -2,13 +2,13 @@
-public class FinalResultDto {
+public class RacingResultResponse {
 
     private final boolean success = true;
```
이건 요구사항에 없는 필드인거 같은데 어디에 쓰이는 건지 알 수 있을까요? 👀

amaran-th replied:
프론트에서 상태코드가 200일 때 정상처리/오류 발생 케이스를 구분해서 처리할 수 있게 하기 위해 추가했습니다!

bimppap replied:
그 뜻은 오류일 때도 200이 간다는 뜻인데, '서버가 요청을 제대로 처리했다'를 의미하는 200을 주면서 오류일 수도 있다는 뜻일까요? 🤔 그럼 프론트 단에서는 응답이 왔을 때 이것이 잘못된 요청인지 아닌지 까보기 전까지는 모를 것 같습니다 🥲

bimppap replied:
Advice 클래스와 연계해서 설명해보자면, 임의로 type을 지정해 어떤 오류인지 알려준다고 했는데 꼭 그럴 필요가 있을까요? HttpStatus만으로도 충분히 어떤 오류인지 파악할 수 있고, 프론트에서도 까보기 전에 status 만으로 요청의 처리 여부를 확인할 수 있을 겁니다.

type와 같은 커스텀 에러코드는 HttpStastus와 에러 메세지만으로는 어디서 발생한 에러인지 정확히 파악하기 어려울 때 도입하는 게 좋습니다 🙂이 점 참고해서 ExceptionResponse의 type과 success 를 없애면서 에러 처리를 할 수 있도록 리팩토링 해보는 건 어떨까요?

amaran-th replied:
상태 코드의 종류를 잘 몰랐는데, 찾아보니 클라이언트의 잘못된 요청을 의미하는 코드가 있네요! 상태코드 400의 응답을 보내도록 했습니다!
처음 type 정보를 넣었던 건 같은 오류 범위에 들어가는 오류들을 세분화하기 위해서였습니다. 예를 들어 동일한 상태코드(400)를 가지지만 빈 값이 입력된 경우는 type 10, 잘못된 유형의 값이 입력된 경우는 type 20, 이런 식으로 오류 코드를 세분화할 수 있을 것 같습니다. 
지금 생각해보니 이건 프로젝트의 룰에 따라 달라질 수 있는 내용이기도 하고, 지금 규모에서는 크게 필요하지 않을 것 같아서 type을 지우고 모든 Exception를 하나로 묶어서 처리하도록 했습니다.
다만 success가 없을 경우 프론트에서 json 데이터 처리 단계에서 올바른 응답(200)과 잘못된 응답(400)을 구분할 수 있는 방법이 없는 것 같아서 그대로 남겨두었습니다...!

bimppap replied:
success가 없어도 가능할텐데, 이건 시도해보셨나요?

```javascript
fetch("/plays", {
      // some code
    })
        .then(response => {
          if (!response.ok) {
            return Promise.reject(response);
          }
          return response.json();
        })
        .then(data => {
            // do something with succeed code
        })
        .catch(error => {
            error.json().then(data => {
              window.alert(data.message);
            })
        });
```

bimppap commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -18,7 +18,7 @@ public class Cars {
     private final NumberGenerator numberGenerator;
 
-    private Cars(List<Car> cars, NumberGenerator numberGenerator) {
+    public Cars(List<Car> cars, NumberGenerator numberGenerator) {
```
흠 과제를 하나 내보죠ㅋㅋ Cars에 있는 static을 전부 없애볼까요?

amaran-th replied:
지웠습니다! Initialize() 메서드의 경우 정적 팩토리 메서드로서 넣은 것이었는데, 정적 팩토리 메서드를 사용하는 것이 좋지 않은 습관인가요??

bimppap replied:
정적 팩토리 메서드를 사용하는게 좋지 않은 습관인 건 아닙니다. 이펙티브 자바에서는 생성자 대신 정적 팩토리 메서드를 쓰라고 권장하기도 하고요.

하지만 정적 팩토리 메서드를 가지고 있으면 상속이 불가능해서 확장성이 떨어지고 다른 정적 메서드(주로 utils라 불리는 패키지에 들어가는 메서드들)와 구분하기 어렵다는 단점도 있습니다.🥲

그런 점에서 정적 패토리 메서드를 사용할 땐 네이밍 컨벤션을 따르는 게 도움이 됩니다. 아니면 다른 개발자들이 헷갈릴 수 있거든요. (제가 바로 이 케이스긴 합니다... 메서드명과 접근제어자만 보고 지나가서 정적 팩토리 메서드가 아닌 그냥 정적 메서드인줄 알았습니다. -ㅡ-;;)

원한다면 원복시키되 메서드명을 바꿔도 좋고 그냥 두셔도 됩니다ㅎㅎ

bimppap replied:
추가로, 아마란스의 리팩토링 속도가 빨라(👍👍) 피드백을 좀 더 드려볼게요. Cars의 책임을 좀 더 작게 쪼개볼 여지가 보입니다. 이건 꼭 반영할 필요는 없지만 (한다면 서비스에서도 개선될 부분이 생길거 같긴 합니다) 한번 생각해보고 얘기를 나눠보면 좋을 것 같아요 🙂

- Car와 Cars처럼 CarName과 CarNames가 있는 건 어떨까요? 
- Cars가 우승자를 계산하는 책임을 가지고 있는 것에 대해 어떻게 생각하세요?

amaran-th replied:
이 정도 규모의 프로젝트에서 CarName을 한 번 더 포장해서 얻을 수 있는 장점이 크게 느껴지지 않습니다. 오히려 코드 양만 늘어나는 게 아닌가요?🤔
우승자를 계산할 때 Cars의 필드에 직접 접근해야 해서 상위 객체(Service)에서 이를 연산하는 건 적절하지 않은 것 같습니다. utils 메서드로 분리해야 한다는 의미이실까요?? 하지만 제가 utils 객체를 구현해본 적이 거의 없어서(이번 미션의 utils 코드는 페어의 레벨1 미션 코드를 가져온 것이라 제가 작성한 게 아닙니다😢) 망설여지네요. 혹시 이에 대한 매뉴얼이 있을까요?

bimppap replied:
이 정도 규모가 작다는 건지 크다는 건지는 모르겠지만...? CarName을 포장해서 얻을 수 있는 장점은 OOP로 설계할 때 얻을 수 있는 장점과 같습니다. OOP를 지향하는 이유는 레벨 1 때 충분히 학습하셨으리라 믿어 따로 서술하진 않겠습니다 🙂

코드를 최대한 짧게 쓰고 싶다면 사실 OOP보단 절차적 프로그래밍을 사용하는 게 더 적절합니다. 하지만 여기선 OOP를 배우고 있는 만큼 최대한 고민해보고 시도해보면 좋을 것 같습니다. 실무에서는 심도 깊게 고민할 시간을 가지기 어렵거든요 🥲

우승자의 계산을 Service에서 연산하는 게 적절하지 않다는 건 공감합니다. Service의 역할은 DAO가 DB에서 받아온 데이터를 전달받아 가공하는 거지만, 직접 하기보단 도메인들에게 메세지를 주고받음으로서 처리해야합니다. 저는 서비스에게 책임을 주기 보다는 게임 결과(우승자 가리기)에 관한 책임을 도맡은 도메인을 따로 만드는 건 어떤가 싶어 물어본 거였습니다. utils 와는 관련 없습니다ㅎㅎ

bimppap commented on src/main/java/racingcar/controller/ConsoleController.java:
```diff
@@ -19,39 +22,33 @@ public ConsoleController(OutputView outputView, InputView inputView) {
     }
 
     public void run() {
```
콘솔컨트롤러 내의 비즈니스로직과 RacingService의 비즈니스 로직의 흐름이 비슷한데, 인터페이스를 활용해서 콘솔 서비스와 웹 서비스로 나누면 콘솔 컨트롤러의 역할이 작아지고 관리하기에도 편하지 않을까요? 👀

amaran-th replied:
반영했습니다!😊

bimppap reviewed (changes requested):
안녕하세요 아마란스, 춘식입니다.
2단계도 빠르게 구현해주셨네요 👍
코멘트 남겼으니 확인 부탁드립니다.
궁금하거나 이해가 안 되는 점은 언제든지 DM 주세요 🙂

Commit:
feat : Controller, Service 기능 개편

- 변수명 수정
- RacingService를 인터페이스로 변경하고 구현체 ConsoleService, WebService 구현
- RacingGame 제거

Commit:
refactor : 생성자 체이닝 적용

Commit:
feat : CarInfoEntity의 패키지 변경 및 id 필드 추가

Commit:
feat : 정적 팩토리 메서드 제거 및 사용하지 않는 메서드 제거

Commit:
refactor DAO 병합 및 명칭 변경

bimppap commented on src/main/java/racingcar/dao/InsertingDAO.java:
```diff
@@ -18,14 +17,13 @@ public InsertingDAO(JdbcTemplate jdbcTemplate) {
 
-    public int insert(Trial trial) {
-        String sql = "INSERT INTO racing (trial_count) values (?)";
+    public int insert() {
```
InsertDao는 이제 사용하지 않으니 없어져도 되겠네요 😎

amaran-th replied:
헉 깜빡했네요ㅋㅋㅋ 삭제했습니다!

bimppap commented on src/main/java/racingcar/entity/CarInfoEntity.java:
```diff
@@ -1,19 +1,25 @@
 
-    public CarInfoDto(int racingId, String name, int position, boolean isWinner) {
+    public CarInfoEntity(int id, int racingId, String name, int position, boolean isWinner) {
+        this.id=id;
```
코드 포맷팅을 습관화합시다!

amaran-th replied:
적용했습니다!

bimppap reviewed (changes requested):
안녕하세요 아마란스, 춘식입니다!
피드백 반영이 정말 빠르네요⚡️⚡️ 리팩토링 잘 해주셨어요.
코멘트 좀 더 남겨봤는데 확인 부탁드립니다.
궁금하거나 이해가 안 되는 부분은 편하게 DM 주세요 👍

Commit:
refactor : 코드 포맷팅 적용 및 불필요한 클래스 제거

Commit:
refactor : type 필드 제거 및 상태 코드 수정

bimppap reviewed (approved):
안녕하세요 아마란스, 춘식입니다!
리팩토링 확인했습니다! 이번 단계는 여기서 마무리해도 될 것 같아 머지하도록 하겠습니다.
질문에 대한 코멘트도 달았으니 확인 바랍니다.
수고 많으셨고 다음 미션도 파이팅입니다! 👍
궁금하거나 대화 나누고 싶은 부분이 있다면 DM 편하게 주세요 🙂
